### PR TITLE
[flink] Optimization of Canal parsing logic

### PR DESCRIPTION
**Pull request by MonsterChenzhuo**
### Purpose
1. Avoid duplicate code

2. Use predefined constants

3. Decomposition big method

### CODE REVIEW

**Comment by MonsterChenzhuo** [comment]
@yuzelin  PTAL,thanks.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java [hunk 1]
```diff
@@ -70,77 +85,42 @@ public CanalRecordParser(
...
-                    after = caseSensitive ? after : keyCaseInsensitive(after);
-                    records.add(
-                            new RichCdcMultiplexRecord(
-                                    databaseName,
-                                    tableName,
-                                    paimonFieldTypes,
-                                    primaryKeys,
-                                    new CdcRecord(RowKind.INSERT, after)));
-                }
+                handleUpdateOperation(
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java:104]: Let the `handleXXXOperation` return the records; then just return in each branch.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java [hunk 2]
```diff
@@ -163,7 +164,7 @@ protected String extractString(String key) {
     }
 
     private void extractDatabaseAndTableNames() {
-        JsonNode tableNode = root.get(FIELD_TABLE);
+        JsonNode tableNode = JsonSerdeUtil.getNodeAs(root, FIELD_TABLE, JsonNode.class);
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java:167]: As commented above.

**Review by yuzelin — COMMENTED** [review]
Left some comments~

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java [hunk 3]
```diff
@@ -27,27 +27,42 @@
...
+ *
+ * <p>The class handles different types of database operations such as INSERT, UPDATE, and DELETE,
+ * and generates corresponding {@link RichCdcMultiplexRecord} objects for each operation.
+ *
+ * <p>Additionally, the parser supports schema extraction, which can be used to understand the
+ * structure of the incoming data and its corresponding field types.
+ */
 public class CanalRecordParser extends RecordParser {
 
+    private static final Charset CHARSET_ISO_8859_1 = StandardCharsets.ISO_8859_1;
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java:64]: done.

[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java:64]: done.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java [hunk 4]
```diff
@@ -27,27 +27,42 @@
...
+    private static final String IS_DDL_FIELD = "isDdl";
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java:65]: done.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java [hunk 5]
```diff
@@ -199,69 +179,67 @@ protected String extractString(String key) {
     }
 
     private boolean isDdl() {
-        return root.get("isDdl") != null && root.get("isDdl").asBoolean();
+        JsonNode isDdlNode = root.get(IS_DDL_FIELD);
+        if (isDdlNode == null) {
+            throw new IllegalArgumentException("Expected 'isDdl' field in the JSON but not found.");
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java:184]: done.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java [hunk 6]
```diff
@@ -27,27 +27,42 @@
...
+ * and generates corresponding {@link RichCdcMultiplexRecord} objects for each operation.
+ *
+ * <p>Additionally, the parser supports schema extraction, which can be used to understand the
+ * structure of the incoming data and its corresponding field types.
+ */
 public class CanalRecordParser extends RecordParser {
 
+    private static final Charset CHARSET_ISO_8859_1 = StandardCharsets.ISO_8859_1;
+    private static final String IS_DDL_FIELD = "isDdl";
     private static final String FIELD_SQL = "sql";
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java:66]: It is used in ValidateFormat,
         if (isDdl()) { // Call the isDdl() method first, and there will be a validity of the ddl field
         // Verify the existence of sql fields
            checkNotNull(root.get(FIELD_SQL), errorMessageTemplate, FIELD_SQL);
        } else {
            checkNotNull(root.get(FIELD_MYSQL_TYPE), errorMessageTemplate, FIELD_MYSQL_TYPE);
            checkNotNull(root.get(FIELD_PRIMARY_KEYS), errorMessageTemplate, FIELD_PRIMARY_KEYS);
        }

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (yuzelin) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java:184]: I mean, we can just check that this field is existed in `validateFormat `, and we don't check nullable here.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java [hunk 7]
```diff
@@ -199,69 +179,67 @@ protected String extractString(String key) {
...
+     * @param record The JSON node containing the data.
+     * @param mySqlFieldTypes A map of MySQL field types.
+     * @param paimonFieldTypes A map of Paimon field types.
+     * @return A map of extracted and transformed data.
+     */
+    private Map<String, String> extractRowFromJson(
             JsonNode record,
             Map<String, String> mySqlFieldTypes,
             LinkedHashMap<String, DataType> paimonFieldTypes) {
+
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java:223]: done.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java [hunk 8]
```diff
@@ -199,69 +179,67 @@ protected String extractString(String key) {
...
-        pkNames.iterator().forEachRemaining(pk -> primaryKeys.add(toFieldName(pk.asText())));
-        return primaryKeys;
+        ArrayNode pkNames = JsonSerdeUtil.getNodeAs(root, FIELD_PRIMARY_KEYS, ArrayNode.class);
+        return StreamSupport.stream(pkNames.spliterator(), false)
+                .map(pk -> toFieldName(pk.asText()))
+                .collect(Collectors.toList());
     }
 
     private LinkedHashMap<String, String> extractFieldTypesFromMySqlType() {
+        JsonNode schema = JsonSerdeUtil.getNodeAs(root, FIELD_MYSQL_TYPE, JsonNode.class);
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java:197]: 1.Direct Retrieval Method:
```java
JsonNode tableNode = root.get(FIELD_TABLE);
```
This is a straightforward method that directly retrieves the child node named FIELD_TABLE from the root node. If the child node doesn't exist, tableNode will be null.
2.Generic Method:
```java
public static <T extends JsonNode> T getNodeAs(
         JsonNode root, String fieldName, Class<T> clazz) {
     ...
}
```
This method is more complex and aims to ensure that the retrieved child node is not only present but also of the expected type. This is achieved through the following steps:

Using Optional to handle potential null values.
Using the map method to retrieve the child node from the root.
Using the filter method to ensure the child node is of the expected type.
Using the map method to cast the child node to the expected type.
If the child node is absent or not of the expected type, an exception is thrown using orElseThrow.


Differences:

Simplicity: The direct retrieval method is simpler but doesn't perform type checks.
Type Safety: The generic method offers type safety, ensuring the retrieved node is of the expected type.
Exception Handling: The generic method throws an exception if the node is absent or not of the expected type, whereas the direct retrieval method returns null if the node is absent.

In summary, the generic method offers stronger type safety and error handling but is also more complex. If you frequently perform such operations in your application and want to reduce potential errors, using the generic method might be a good choice.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java [hunk 9]
```diff
@@ -199,69 +179,67 @@ protected String extractString(String key) {
...
-                }
-            }
-            resultMap.put(fieldName, newValue);
-        }
+        Map<String, String> resultMap =
+                mySqlFieldTypes.entrySet().stream()
+                        .filter(
+                                entry ->
+                                        jsonMap.containsKey(entry.getKey())
+                                                && jsonMap.get(entry.getKey()) != null)
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java:235]: done.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java [hunk 10]
```diff
@@ -273,4 +251,99 @@ private Map<String, String> extractRow(
...
+                        CanalFieldParser.convertGeoType2WkbArray(
+                                oldValue.getBytes(CHARSET_ISO_8859_1));
+                newValue = MySqlTypeUtils.convertWkbArray(wkb);
+            } catch (Exception e) {
+                throw new IllegalArgumentException(
+                        String.format("Failed to convert %s to geometry JSON.", oldValue), e);
+            }
+        }
+        return newValue;
+    }
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java:275]: done.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java (hunk 1 above)
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java:104]: done.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/canal/CanalRecordParser.java]: done.

**Comment by MonsterChenzhuo** [comment]
@yuzelin PTAL,thanks.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java (hunk 2 above)
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java:167]: done.

**Review by yuzelin — APPROVED** [review]
+1
